packages/talos_agent/src/agentv2/agent_v2.rs

```diff
@@ -0,0 +1,58 @@
+use crate::agentv2::model::{CancelRequestChannelMessage, CertifyRequestChannelMessage};
+use crate::agentv2::state_manager::StateManager;
+use crate::api::{AgentConfig, CertificationRequest, CertificationResponse, KafkaConfig, TalosAgent, TalosIntegrationType};
+use async_trait::async_trait;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use time::OffsetDateTime;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+pub struct TalosAgentImplV2 {
+    rx_certify: Sender<CertifyRequestChannelMessage>,
+    _rx_cancel: Sender<CancelRequestChannelMessage>,
+}
+
+impl TalosAgentImplV2 {
+    pub async fn new(
+        agent_config: AgentConfig,
+        kafka_config: Option<KafkaConfig>,
+        int_type: &TalosIntegrationType,
+        certify: (Sender<CertifyRequestChannelMessage>, Receiver<CertifyRequestChannelMessage>),
+        cancel: (Sender<CancelRequestChannelMessage>, Receiver<CancelRequestChannelMessage>),
+        publish_times: &Arc<Mutex<HashMap<String, u64>>>,
+    ) -> Result<TalosAgentImplV2, String> {
+        let state_manager = StateManager::new(agent_config, kafka_config, int_type, publish_times).unwrap();
+
+        state_manager.start(certify.1, cancel.1).await;
+
+        Ok(TalosAgentImplV2 {
+            rx_certify: certify.0,
+            _rx_cancel: cancel.0,
+        })
+    }
+}
+
+#[async_trait]
+impl TalosAgent for TalosAgentImplV2 {
+    async fn certify(&self, request: CertificationRequest) -> Result<CertificationResponse, String> {
+        let (tx, mut rx) = mpsc::channel::<CertificationResponse>(1);
+
+        let m = CertifyRequestChannelMessage::new(&request, &tx);
+
+        let to_state_manager = self.rx_certify.clone();
+        let response = match to_state_manager.send(m).await {
+            Ok(()) => match rx.recv().await {
```

Comment:
I understand why you do this here from performance testing perspective. But in real world scenario this should be avoided. There is no guarantee that we are receiving the response of the message we send. It could be some other message. 

Why you don't notice this problem in performance testing is because you run a for loop of tasks which are run serially.

Reply:
Hold on. This channel instance is for the specific request. The only message we will receive through it is the reply to our request, no other reply can come to this channel instance.

Reply:
This is related to [this comment ](https://github.com/kindredgroup/talos/pull/10#discussion_r1116552094) I just made. your perf test is currently running 1 task at a time. And that is why this is fine. Ideally when agent publishes in parallel, there is no guarantee that the certifier will return the decisions in the same order the candidate messages were received.

Reply:
`mpsc` can have only 1 receiver. Since you create only one instance of `agent` and then put `agent` in `Arc` it's basically the same receiver being used across threads. Could you check and confirm on this?

Reply:
Are we talking about lines 43 (variable "rx") and line 49 (variable "rx"). Its new every time someone calls certify()

Reply:
ignore this comment as well, I was looking at the wrong `receiver` all along :(